state-chain/runtime/src/lib.rs

```diff
@@ -2094,12 +2106,107 @@ impl_runtime_apis! {
 		}
 
 		fn cf_validate_dca_params(number_of_chunks: u32, chunk_interval: u32) -> Result<(), DispatchErrorWithMessage> {
-			pallet_cf_swapping::Pallet::<Runtime>::validate_dca_params(&cf_primitives::DcaParameters{number_of_chunks, chunk_interval}).map_err(Into::into)
+			pallet_cf_swapping::Pallet::<Runtime>::validate_dca_params(&DcaParameters{number_of_chunks, chunk_interval}).map_err(Into::into)
 		}
 
 		fn cf_validate_refund_params(retry_duration: u32) -> Result<(), DispatchErrorWithMessage> {
 			pallet_cf_swapping::Pallet::<Runtime>::validate_refund_params(retry_duration).map_err(Into::into)
 		}
+
+		fn cf_get_vault_swap_details(
+			_broker: AccountId,
+			source_asset: Asset,
+			destination_asset: Asset,
+			destination_address: ForeignChainAddress,
+			broker_commission: BasisPoints,
+			min_output_amount: AssetAmount,
+			retry_duration: u32,
+			boost_fee: Option<BasisPoints>,
+			affiliate_fees: Option<Affiliates<AccountId>>,
+			dca_parameters: Option<DcaParameters>,
+		) -> Result<VaultSwapDetails, DispatchErrorWithMessage> {
+			// Validate params
+			if broker_commission != 0 || affiliate_fees.map_or(false, |fees| !fees.is_empty()) {
+				return Err(Into::<DispatchErrorWithMessage>::into(DispatchError::from(
+					pallet_cf_swapping::Error::<Runtime>::VaultSwapBrokerFeesNotSupported,
+				)));
+			}
+			pallet_cf_swapping::Pallet::<Runtime>::validate_refund_params(retry_duration)
+				.map_err(Into::<DispatchErrorWithMessage>::into)?;
+			if let Some(params) = dca_parameters.as_ref() {
+				pallet_cf_swapping::Pallet::<Runtime>::validate_dca_params(params)
+					.map_err(Into::<DispatchErrorWithMessage>::into)?;
+			}
+			if ForeignChain::from(destination_asset) != destination_address.chain() {
+				return Err(Into::<DispatchErrorWithMessage>::into(DispatchError::from(
+					pallet_cf_swapping::Error::<Runtime>::IncompatibleDestinationAddress,
+				)));
+			}
+
+			// Encode swap
+			match ForeignChain::from(source_asset) {
+				ForeignChain::Bitcoin => {
+					let params = UtxoEncodedData {
+						output_asset: destination_asset,
+						output_address: to_encoded_address(
+							destination_address,
+							Environment::network_environment,
+						),
+						parameters: SharedCfParameters {
+							retry_duration: retry_duration.try_into().map_err(|_| {
+								Into::<DispatchErrorWithMessage>::into(DispatchError::from(
+									pallet_cf_swapping::Error::<Runtime>::SwapRequestDurationTooLong,
+								))
+							})?,
+							min_output_amount,
+							number_of_chunks: dca_parameters
+								.as_ref()
+								.map(|params| params.number_of_chunks)
+								.unwrap_or(1)
+								.try_into()
+								.map_err(|_| {
+									Into::<DispatchErrorWithMessage>::into(DispatchError::from(
```

Comment:
These error conversions are quite repetitive - maybe we just need to write the correct `From` implementation for DispatchErrorWithMessage? 

At the very least we could write a struct method to make this more readable. Something like this might do the trick:

```
impl DispatchErrorWithMessage {
	pub fn from_pallet_error(e: impl Into<DispatchError>) {
		Self::from(e.into())
	}
}
```